sdk/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleSummaryAggregator.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics.aggregator;
+
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.opentelemetry.sdk.metrics.data.MetricData.DoubleSummaryPoint;
+import io.opentelemetry.sdk.metrics.data.MetricData.Point;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+@ThreadSafe
+public class DoubleSummaryAggregator extends AbstractAggregator {
+  private static final AggregatorFactory AGGREGATOR_FACTORY =
+      new AggregatorFactory() {
+        @Override
+        public Aggregator getAggregator() {
+          return new DoubleSummaryAggregator();
+        }
+      };
+
+  // The current value. This controls its own internal thread-safety via method access. Don't
+  // try to use its fields directly.
+  private final DoubleSummary current = new DoubleSummary();
+
+  public static AggregatorFactory getFactory() {
+    return AGGREGATOR_FACTORY;
+  }
+
+  @Override
+  void doMergeAndReset(Aggregator target) {
+    DoubleSummaryAggregator other = (DoubleSummaryAggregator) target;
+
+    DoubleSummary copy = current.copyAndReset();
+    other.current.update(copy.count, copy.sum, copy.min, copy.max);
+  }
+
+  @Nullable
+  @Override
+  public Point toPoint(long startEpochNanos, long epochNanos, Map<String, String> labels) {
+    return current.toPoint(startEpochNanos, epochNanos, labels);
+  }
+
+  @Override
+  public void recordDouble(double value) {
+    current.record(value);
+  }
+
+  private static class DoubleSummary {
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
I'll try both and see if there's any performance difference. I like the clear semantics of readers & writers, but if it makes the perf worse, then that's a really good reason to change it. I'll try both and see how it looks from a benchmarking perspective.

Reply:
I ran benchmarks on both java 8 and java 11, and for both of them, the regular ReentrantLock version was slightly slower than the ReadWrite lock version. So, I vote we stay with the R/W lock because it appears to provide a small amount of performance gain, and makes the semantics of the updates/reads slightly more clear.

Reply:
I can confirm. In my runs, ReadWrite lock is roughly 1%, 5%, and 10% faster with 1, 5, and 10 threads, respectively.

Reply:
FYI, I also tried re-implementing this with 4 atomics, uncoordinated, and it was significantly slower than the locking version from the benchmarks. This surprised me, but that's why we benchmark!